src/recipes.jl

```diff
@@ -378,109 +378,256 @@ end
-    counts = zeros(length(edges)-1)
+@recipe function f(::Type{Val{:stepbins}}, x, y, z)
+    edge, weights = x, y
+
+    axis = d[:subplot][Plots.isvertical(d) ? :xaxis : :yaxis]
 
-    # add a weighted count
-    for (i,vi) in enumerate(v)
-        idx = bucket_index(vi, edges)
-        counts[idx] += (weights == nothing ? 1.0 : weights[i])
+    xpts, ypts = _stepbins_path(edge, weights)
+    if !Plots.isvertical(d)
+        xpts, ypts = ypts, xpts
     end
 
-    counts = isapprox(extrema(diff(edges))...) ? counts : counts ./ diff(edges) # for uneven bins, normalize to area.
+    # create a secondary series for the markers
+    if d[:markershape] != :none
+        @series begin
+            seriestype := :scatter
+            x := Plots._bin_centers(edge)
+            y := weights
+            fillrange := nothing
+            label := ""
+            primary := false
+            ()
+        end
+        markershape := :none
+        xerror := :none
+        yerror := :none
+    end
 
-    # normalize by bar area?
-    norm_denom = normed ? sum(diff(edges) .* counts) : 1.0
-    if norm_denom == 0
-        norm_denom = 1.0
+    x := xpts
+    y := ypts
+    seriestype := :path
+    ylims --> [0, 1.1 * maximum(weights)]
+    ()
+end
+Plots.@deps stepbins path
+
+
+function _auto_binning_nbins{N}(vs::NTuple{N,AbstractVector}, dim::Integer; mode::Symbol = :auto)
+    _cl(x) = max(ceil(Int, x), 1)
+    _iqr(v) = quantile(v, 0.75) - quantile(v, 0.25)
+    _span(v) = maximum(v) - minimum(v)
+
+    n_samples = length(linearindices(first(vs)))
+    # Estimator for number of samples in one row/column of bins along each axis:
+    n = max(1, n_samples^(1/N))
+
+    v = vs[dim]
+
+    if mode == :sqrt  # Square-root choice
+        _cl(sqrt(n))
+    elseif mode == :sturges || mode ==:auto  # Sturges' formula
+        _cl(log2(n)) + 1
+    elseif mode == :rice  # Rice Rule
+        _cl(2 * n^(1/3))
+    elseif mode == :scott  # Scott's normal reference rule
+        _cl(_span(v) / (3.5 * std(v) / n^(1/3)))
+    elseif mode == :fd  # Freedman–Diaconis rule
+        _cl(_span(v) / (2 * _iqr(v) / n^(1/3)))
+    else
+        error("Unknown auto-binning mode $mode")
     end
+end
+
+_hist_edge{N}(vs::NTuple{N,AbstractVector}, dim::Integer, binning::Integer) = StatsBase.histrange(vs[dim], binning, :left)
+_hist_edge{N}(vs::NTuple{N,AbstractVector}, dim::Integer, binning::Symbol) = _hist_edge(vs, dim, _auto_binning_nbins(vs, dim, mode = binning))
+_hist_edge{N}(vs::NTuple{N,AbstractVector}, dim::Integer, binning::AbstractVector) = binning
 
-    edges, counts ./ norm_denom
+_hist_edges{N}(vs::NTuple{N,AbstractVector}, binning::NTuple{N}) =
+    map(dim -> _hist_edge(vs, dim, binning[dim]), (1:N...))
+
+_hist_edges{N}(vs::NTuple{N,AbstractVector}, binning::Union{Integer, Symbol, AbstractVector}) =
+    map(dim -> _hist_edge(vs, dim, binning), (1:N...))
+
+_hist_norm_mode(mode::Symbol) = mode
+_hist_norm_mode(mode::Bool) = mode ? :norm : :none
```

Comment:
> I don't think we should be locked to doing what StatsBase does.

I'd very much prefer to conform with StatsBase, but I could live with a different default if we really must do things differently for Plots - as long as it's clearly marked. @tbreloff ?

> I also don't agree with your last point - Plots should apply the :density normalization for inhomogenous bins and mark it clearly in the ylabels.

I'm sorry, but I have to vehemently disagree here. For once the meaning of the plot (and it would be changed) shouldn't depend on axis binning - such auto-magical things tend to be very intransparent to the user, and hence very dangerous. People may spend hours trying to find out why they get counts in some plot and a density on another, just because the edges are different (users tend not to read the full docs first, esp. since Plots is so easy to use). You also can't automatically produce correct y-labels, because Plots can't know what the x-axis units are. And even if you could, automatically setting ylabels and changing y-units depending on how your edges are structured is just, uh, weird.

2d-histgrams are even more problematic - just looking at the colors, the user may never realize that he's not plotting what he thought he was (and where do you put the automatic z-axis label, even if you'd know what to put in it?). And that if just one axis is inhomogenous? In any case, such non-obvious automatic behavior changes could, in the worst case, lead to incorrect scientific results or interpretations.
 
 I stand by my point - inhomogenous axes don't happen by accident, and they're not that common. People who use them will usually know what they want - having plots suddenly change the meaning of the bin values will make their life harder, not easier. Such histograms simply have more complex behaviour than homogenously-binned ones - trying to "hide" that mathematical reality from the users won't do them any favours.